Round response cost to 6 decimal places in headers

colesmcintosh opened the pull request:
## Title

Round response cost to 6 decimal places in headers

## Relevant issues

https://github.com/BerriAI/litellm/issues/8867

## Pre-Submission checklist

- [x] I have Added testing in the `tests/litellm/` directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on `make test-unit` (https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

🆕 New Feature

## Changes

This PR improves the readability of response cost values in HTTP headers by rounding them to 6 decimal places. Previously, the response cost was displayed with full floating-point precision, which could be difficult to read. Now, the value is rounded to 6 decimal places for better readability while maintaining sufficient precision for cost tracking.

Additionally, this PR fixes a typo in the cache parameter name from `default_in_redis_ttl` to `default_redis_ttl` to match the actual parameter name used throughout the codebase.

colesmcintosh committed:
fix(proxy): correct response cost rounding and Redis TTL config

colesmcintosh committed:
test(proxy): add test for response cost rounding in headers

ishaan-jaff commented on litellm/proxy/proxy_server.py:
```diff
@@ -806,7 +806,7 @@ def get_custom_headers(
         "x-litellm-version": version,
         "x-litellm-model-region": model_region,
-        "x-litellm-response-cost": str(response_cost),
+        "x-litellm-response-cost": str(round(response_cost, 6)),
```
use a helper static method here please, can it be a safe method (if it fails it still returns str(response cost) 

my concern is getting the rounded response cost fails, the llm response should not fail

ishaan-jaff commented on litellm/proxy/proxy_server.py:
```diff
@@ -1638,7 +1638,7 @@ def _init_cache(
 
         if "default_redis_ttl" in cache_params:
-            litellm.default_redis_ttl = cache_params["default_in_redis_ttl"]
+            litellm.default_redis_ttl = cache_params["default_redis_ttl"]
```
nice catch

ishaan-jaff commented on tests/proxy_unit_tests/test_proxy_server.py:
```diff
@@ -2208,3 +2208,34 @@ def test_get_timeout_from_request():
     timeout = LiteLLMProxyRequestSetup._get_timeout_from_request(headers)
     assert timeout == 90.5
+
+
```
can you add this in tests/litellm/proxy/test_proxy_server.py 

we are moving to organizing our tests in that manner for contributor PRs

ishaan-jaff reviewed (changes requested):
minor changes needed

colesmcintosh committed:
feat(proxy): add safe response cost rounding function

Introduces a new utility function `_safely_round_response_cost()` to handle response cost rounding more robustly, preventing potential errors when rounding or converting cost values in headers.

colesmcintosh committed:
test(proxy): expand test cases for response cost rounding

colesmcintosh committed:
test(proxy): move test_proxy_server.py into tests/litellm/proxy

ishaan-jaff commented on litellm/proxy/proxy_server.py:
```diff
@@ -781,6 +781,21 @@ async def redirect_ui_middleware(request: Request, call_next):
+    except Exception:
+        # If rounding fails for any reason, return the original value
+        return str(response_cost)
+
```
move this to proxy/utils.py

colesmcintosh replied:
@ishaan-jaff should this removed from `litellm/proxy/utils.py` as it's being handled in `litellm/proxy/common_request_processing.py` as a static method on ProxyBaseLLMRequestProcessing?

ishaan-jaff commented on tests/litellm/proxy/test_proxy_server.py:
```diff
@@ -0,0 +1,2239 @@
+# Run the test
+# test_openai_deployments_model_chat_completions_azure()
+
+
```
for this PR can you just add your new test to this file, I want to make sure it lands into main

ishaan-jaff reviewed (changes requested):
left review

colesmcintosh committed:
Merge remote-tracking branch 'origin/main' into 8867-x-litellm-response-cost-header-rounding

colesmcintosh committed:
Merge remote-tracking branch 'origin/main' into 8867-x-litellm-response-cost-header-rounding

colesmcintosh committed:
refactor(proxy): move response cost rounding function to utils

ishaan-jaff commented on litellm/proxy/proxy_server.py:
```diff
@@ -781,33 +781,160 @@ async def redirect_ui_middleware(request: Request, call_next):
+        "x-litellm-key-tpm-limit": str(user_api_key_dict.tpm_limit),
+        "x-litellm-key-rpm-limit": str(user_api_key_dict.rpm_limit),
+        "x-litellm-key-max-budget": str(user_api_key_dict.max_budget),
+        "x-litellm-key-spend": str(user_api_key_dict.spend),
```
can you use this for `x-litellm-key-spend` + `x-litellm-key-max-budget` too

ishaan-jaff commented on litellm/proxy/proxy_server.py:
```diff
@@ -5157,7 +5283,7 @@ async def run_thread(
         api_base = hidden_params.get("api_base", None) or ""
 
         fastapi_response.headers.update(
-            ProxyBaseLLMRequestProcessing.get_custom_headers(
```
can you not undo this change, we recently moved this to ProxyBaseLLMRequestProcessing

ishaan-jaff commented on litellm/proxy/utils.py:
```diff
@@ -115,6 +115,21 @@ def safe_deep_copy(data):
 
 
+def _safely_round_response_cost(response_cost: Optional[Union[float, str]]) -> str:
+    """
```
this should be a static method in ProxyBaseLLMRequestProcessing

ishaan-jaff commented on tests/litellm/proxy/test_proxy_server.py:
```diff
@@ -0,0 +1,2239 @@
+    )
+    print(f"resp: {resp}")
+    models = resp["data"]
+    is_model_alias_in_list = False
```
you just need your new response cost test in this file, don't copy over everything else

ishaan-jaff commented on litellm/proxy/proxy_server.py:
```diff
@@ -449,7 +449,7 @@ async def proxy_startup_event(app: FastAPI):
     ## CHECK PREMIUM USER
     verbose_proxy_logger.debug(
-        "litellm.proxy.proxy_server.py::startup() - CHECKING PREMIUM USER - {}".format(
+        "litellm/proxy/proxy_server.py::startup() - CHECKING PREMIUM USER - {}".format(
```
why this ?

colesmcintosh replied:
This change was unintentional—I accidentally included it and will revert

ishaan-jaff reviewed (changes requested):
reviewed

colesmcintosh committed:
Merge remote-tracking branch 'upstream/main' into 8867-x-litellm-response-cost-header-rounding

colesmcintosh committed:
fix(proxy): revert path notation from slash to dot

colesmcintosh committed:
refactor(tests): add only test_response_cost_rounding to test_proxy_server.py

colesmcintosh committed:
refactor(proxy): update header retrieval to use ProxyBaseLLMRequestProcessing

colesmcintosh committed:
feat(proxy): add static method to safely round response cost to six decimal places

colesmcintosh committed:
refactor(proxy): use static method for rounding response cost

colesmcintosh committed:
Merge remote-tracking branch 'upstream/main' into 8867-x-litellm-response-cost-header-rounding

colesmcintosh committed:
fix(tests): update import path for response cost rounding test

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/BerriAI/litellm?pullRequest=9118) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/BerriAI/litellm?pullRequest=9118) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/BerriAI/litellm?pullRequest=9118) it.</sub>

colesmcintosh committed:
Merge main branch into 8867-x-litellm-response-cost-header-rounding
